pkg/server/sandbox_status.go

```diff
@@ -17,14 +17,96 @@ limitations under the License.
 package server
 
 import (
-	"errors"
+	"fmt"
 
+	"github.com/golang/glog"
 	"golang.org/x/net/context"
 
+	"github.com/containerd/containerd/api/services/execution"
+	"github.com/containerd/containerd/api/types/container"
+
 	"k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+
+	"github.com/kubernetes-incubator/cri-containerd/pkg/metadata"
 )
 
 // PodSandboxStatus returns the status of the PodSandbox.
-func (c *criContainerdService) PodSandboxStatus(ctx context.Context, r *runtime.PodSandboxStatusRequest) (*runtime.PodSandboxStatusResponse, error) {
-	return nil, errors.New("not implemented")
+func (c *criContainerdService) PodSandboxStatus(ctx context.Context, r *runtime.PodSandboxStatusRequest) (retRes *runtime.PodSandboxStatusResponse, retErr error) {
+	glog.V(4).Infof("PodSandboxStatus for sandbox %q", r.GetPodSandboxId())
+	defer func() {
+		if retErr == nil {
+			glog.V(4).Infof("PodSandboxStatus returns status %+v", retRes.GetStatus())
+		}
+	}()
+
+	// Convert potential truncated id to full id.
+	id, err := c.sandboxIDIndex.Get(r.GetPodSandboxId())
+	if err != nil {
+		return nil, fmt.Errorf("failed to find sandbox id %q: %v", r.GetPodSandboxId(), err)
+	}
+
+	sandboxInStore, err := c.sandboxStore.Get(id)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get sandbox metadata from store for %q: %v",
+			id, err)
+	}
+	// TODO(random-liu): Consolidate the behavior of ImageStatus and PodSandboxStatus
+	// in CRI.
+	if sandboxInStore == nil {
+		return nil, fmt.Errorf("sandbox %q does not exist", id)
+	}
+	meta, err := sandboxInStore.GetMetadata()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get sandbox metadata for %q: %v", id, err)
+	}
+
+	info, err := c.containerService.Info(ctx, &execution.InfoRequest{ID: id})
+	if err != nil && !isContainerdContainerNotExistError(err) {
+		return nil, fmt.Errorf("failed to get sandbox container info for %q: %v", id, err)
+	}
+
+	// Set sandbox state to NOTREADY by default.
+	state := runtime.PodSandboxState_SANDBOX_NOTREADY
+	// If the sandbox container is running, just return as it is.
+	if info != nil && info.Status == container.Status_RUNNING {
+		state = runtime.PodSandboxState_SANDBOX_READY
```

Comment:
same as above.. where's the original container state... why only ready or not ready...